### PR TITLE
Revert lru_cache to functional form; Add lru-dict for integrators

### LINKED ISSUE #1442: Excessive RAM consumption when running multiple simulations (Solver Issue)
# Problem description
When using the same solver object for multiple simulations, memory consumption increases every simulation. The issue is triggered by the solver object. (If this is desired behaviour, please close the issue.)

[Link to Slack Discussion](https://pybamm.slack.com/archives/C010LMK2ZB9/p1616409488024100)

### This minimal example triggers the Issue
```python
import pybamm as pb
import gc
​
solver = pb.CasadiSolver(mode="safe")
para = pb.ParameterValues(chemistry=pb.parameter_sets.Chen2020)
model = pb.lithium_ion.SPM()
exp = pb.Experiment(["Discharge at 0.2C until 2.9V (1 minutes period)"])
​
for i in range(30):
    print(i)
    sim = pb.Simulation(model, experiment=exp, parameter_values=para, solver=solver)
    sim.solve()
​
    del sim
    gc.collect()
```

### This minimal example is a workaround for the Issue
```python
import pybamm as pb
import gc
​
para = pb.ParameterValues(chemistry=pb.parameter_sets.Chen2020)
exp = pb.Experiment(["Discharge at 0.2C until 2.9V (1 minutes period)"])
​
model = pb.lithium_ion.SPM()
​
​
for i in range(30):
    print(i)
    solver = pb.CasadiSolver(mode="safe") # creating solver object every iteration
    sim = pb.Simulation(model, experiment=exp, parameter_values=para, solver=solver)
    sim.solve()
​
    del sim, solver # deleting object every iteration frees the space
    gc.collect()
```

### PR DESCRIPTION
**Pull request by jsbrittain**
# Description

Fix for excessive RAM consumption when running back-to-back simulations. As outlined in #1442, three problems were identified:
1) Use of the decorator @lru_cache on methods prevent the class instance from being garbage collected. Declaring lru_cache methods functionally in the initialiser resolves this problem. However, note that the function is no longer picklable, and instead the cache relationship must be removed and reinstated during pickling / unpickling.
2) Integrators are stored in a dictionary for re-use in CasadiSolver. Successive simulations with the same solver can cause the Solver object to grow unbound. A new LRU dictionary (LRUDict) has been introduced to limit the total number of items stored (along with appropriate test code). Default behaviour remains unchanged, but user-specification of an integrator limit permits simulations to run in a loop without unbound RAM consumption. Documentation has been updated to reflect the new parameter.
3) Successive caching of Solution histories, which can be solved by passing `solution.last_state` to Simulation.solve() (but requires resolution of #2788).

Fixes #1442 (Excessive RAM consumption when running multiple simulations)

## Type of change

Please add a line in the relevant section of [CHANGELOG.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CHANGELOG.md) to document the change (include PR #) - note reverse order of PR #s. If necessary, also add to the list of breaking changes.

- [X] New feature (non-breaking change which adds functionality)
- [ ] Optimization (back-end change that speeds up the code)
- [X] Bug fix (non-breaking change which fixes an issue)

# Key checklist:

- [X] No style issues: `$ pre-commit run` (see [CONTRIBUTING.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CONTRIBUTING.md#installing-and-using-pre-commit) for how to set this up to run automatically when committing locally, in just two lines of code)
- [X] All tests pass: `$ python run-tests.py --all`
- [X] The documentation builds: `$ python run-tests.py --doctest`

You can run unit and doctests together at once, using `$ python run-tests.py --quick`.

## Further checks:

- [X] Code is commented, particularly in hard-to-understand areas
- [X] Tests added that prove fix is effective or that feature works

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -8,13 +8,15 @@
 - Added "Open-circuit voltage [V]", which is the open-circuit voltage as calculated from the bulk particle concentrations. The old variable "Measured open circuit voltage [V]", which referred to the open-circuit potential as calculated from the surface particle concentrations, has been renamed to "Surface open-circuit voltage [V]". ([#2740](https://github.com/pybamm-team/PyBaMM/pull/2740))
 - Added an example for `plot_voltage_components`, explaining what the different voltage components are. ([#2740](https://github.com/pybamm-team/PyBaMM/pull/2740))
 - Added method to calculate maximum theoretical energy. ([#2777](https://github.com/pybamm-team/PyBaMM/pull/2777)) and add to summary variables ([#2781](https://github.com/pybamm-team/PyBaMM/pull/2781))
+- Added option to limit the number of integrators stored in CasadiSolver, which is particularly relevant when running simulations back-to-back [#2823](https://github.com/pybamm-team/PyBaMM/pull/2823)
```
Comment (brosaplanella) [inline CHANGELOG.md:11]: Can you move the comment further up (we follow a decreasing PR number sorting)? I realise some of the existing lines do not follow the order actually, so if you can sort them that would be great hahah.

**Review by brosaplanella — APPROVED** [review]
Looks good, thanks! Just a small comment on the CHANGELOG.

**Comment by brosaplanella** [comment]
@all-contributors add @jsbrittain for code and tests

**Review by martinjrobins — COMMENTED** [review]
thanks @jsbrittain. The one comment I had was the default of "unbounded" for number of stored integrators. What is the effect of having a finite bound by default (e.g. 100?), it feels like it would be nice to have memory consumption bounded by default.

File: CHANGELOG.md (hunk 1 above)
Comment (jsbrittain) [reply CHANGELOG.md:11]: Sure, will do!

**Pull request merged**